src/Components/Web.JS/src/Boot.Web.ts

```diff
@@ -46,56 +58,101 @@ function boot(options?: Partial<WebStartOptions>) : Promise<void> {
 
   attachComponentDescriptorHandler(descriptorHandler);
   attachStreamingRenderingListener(options?.ssr, navigationEnhancementCallbacks);
+  attachAutoModeResolver(resolveAutoMode);
 
   if (!options?.ssr?.disableDomPreservation) {
     attachProgressivelyEnhancedNavigationListener(navigationEnhancementCallbacks);
   }
 
   registerAllComponentDescriptors(document);
+  handleUpdatedComponentDescriptors();
 
   return Promise.resolve();
 }
 
-function registerComponentDescriptor(descriptor: ServerComponentDescriptor | WebAssemblyComponentDescriptor) {
-  switch (descriptor.type) {
-    case 'server':
-      startCircuitIfNotStarted();
-      circuitRootComponents.registerComponentDescriptor(descriptor);
-      break;
-    case 'webassembly':
-      startWebAssemblyIfNotStarted();
-      webAssemblyRootComponents.registerComponentDescriptor(descriptor);
-      break;
+function resolveAutoMode(): 'server' | 'webassembly' | null {
+  if (webAssemblyLoadingState >= WebAssemblyLoadingState.Loaded) {
+    // The WebAssembly runtime has loaded or is actively starting, so we'll use
+    // WebAssembly for the component in question. We'll also start
+    // the WebAssembly runtime if it hasn't already.
+    startWebAssemblyIfNotStarted();
+    return 'webassembly';
+  }
+
+  if (autoModeTimeoutState === 'timed out') {
+    // We've waited too long for WebAssembly to initialize, so we'll use the Server
+    // render mode for the component in question. At some point if the WebAssembly
+    // runtime finishes loading, we'll start using it again due to the earlier
+    // check in this function.
+    startCircuitIfNotStarted();
+    return 'server';
+  }
+
+  if (autoModeTimeoutState === undefined) {
+    // The WebAssembly runtime hasn't loaded yet, and this is the first
+    // time auto mode is being requested.
+    // We'll wait a bit for the WebAssembly runtime to load before making
+    // a render mode decision.
+    autoModeTimeoutState = 'waiting';
+    setTimeout(() => {
+      autoModeTimeoutState = 'timed out';
+
+      // We want to ensure that we activate any markers whose render mode didn't get resolved
+      // earlier.
+      handleUpdatedComponentDescriptors();
+    }, autoModeWebAssemblyTimeoutMilliseconds);
+  }
```

Comment:
As above, I'm assuming it's because we have no choice but to go through this timeout. If I'm wrong and we could simply not do this and just resolve deterministically based on whether the files are already cached, I think that would be a much stronger and clearer starting point.